Remove BUILD_NAMEDTENSOR macros

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31047 Remove BUILD_NAMEDTENSOR from codegen and .cu files
* **#30894 Remove BUILD_NAMEDTENSOR macros**

This PR begins the process of removing BUILD_NAMEDTENSOR macros. There
will be followups.

Reasons for removing the macros:
- BUILD_NAMEDTENSOR is always on and has been on since pytorch 1.3.0.
- Since we don't test building without it, it is useless to keep around.
- Code becomes nicer to read without the macros

Reasons for not removing the macros:
- potential for feature flagging
- binary size reduction / pytorch modularity. I don't think this is a
major consideration because macros like this are not the best way to
achieve this.

Now, I argue against needing to feature flag. The main reason why we
might want to feature flag is if we need to disable the feature.
We'd need a fast switch to disable the feature if someone discovers
in the future that named tensors caused some regression in some existing workflows.

In https://github.com/pytorch/pytorch/pull/25798, I did a variety of
macro- and micro- benchmarks to determine the performance impact of named
tensors on regular tensors.

[The
microbenchmarks](https://github.com/pytorch/pytorch/pull/25798#issuecomment-529014810)
were not very stable, and running the
microbenchmarks for more iterations doesn't actually help because the
noise is not distributed in a nice way. Instead of microbenchmarks I ran
a [profiler
(perf)](https://github.com/pytorch/pytorch/pull/25798#issuecomment-555707645)
to estimate how much overhead named tensors add to unnamed code. I
estimated the overhead to be less than 100ns for `add` and even smaller
for `mm`; there are ways to optimize even futher if we find this to be a
problem.

[Initial
macrobenchmarks](https://github.com/pytorch/pytorch/pull/25798#issuecomment-530539104)
were also not very stable. I ran imagenet for some number of epochs. To
make them more stable, I got rid of the data loading (which seemed to
vary between runs). [In some benchmarkers without data
loading](https://github.com/pytorch/pytorch/pull/25798#issuecomment-562214053),
we can see that the results are less noisy now. These results support
no noticeable regressions in speed.

Test Plan:
- wait for CI

Differential Revision: [D18858543](https://our.internmc.facebook.com/intern/diff/D18858543)

zou3519 committed:
Remove BUILD_NAMEDTENSOR macros

This PR begins the process of removing BUILD_NAMEDTENSOR macros. There
will be followups.

Reasons for removing the macros:
- BUILD_NAMEDTENSOR is always on and has been on since pytorch 1.3.0.
- Since we don't test building without it, it is useless to keep around.
- Code becomes nicer to read without the macros

Reasons for not removing the macros:
- potential for feature flagging

Now, I argue against needing to feature flag. The main reason why we
might want to feature flag is if we need to disable the feature.
We'd need a fast switch to disable the feature if someone discovers
in the future that named tensors caused some regression in some existing workflows.

In https://github.com/pytorch/pytorch/pull/25798, I did a variety of
macro- and micro- benchmarks to determine the performance impact of named
tensors on regular tensors.

[The
microbenchmarks](https://github.com/pytorch/pytorch/pull/25798#issuecomment-529014810)
were not very stable, and running the
microbenchmarks for more iterations doesn't actually help because the
noise is not distributed in a nice way. Instead of microbenchmarks I ran
a [profiler
(perf)](https://github.com/pytorch/pytorch/pull/25798#issuecomment-555707645)
to estimate how much overhead named tensors add to unnamed code. I
estimated the overhead to be less than 100ns for `add` and even smaller
for `mm`; there are ways to optimize even futher if we find this to be a
problem.

[Initial
macrobenchmarks](https://github.com/pytorch/pytorch/pull/25798#issuecomment-530539104)
were also not very stable. I ran imagenet for some number of epochs. To
make them more stable, I got rid of the data loading (which seemed to
vary between runs). [In some benchmarkers without data
loading](https://github.com/pytorch/pytorch/pull/25798#issuecomment-562214053),
we can see that the results are less noisy now. These results support
no noticeable regressions in speed.

Test Plan:
- wait for CI

[ghstack-poisoned]

nairbv commented:
> Reasons for not removing the macros:

One other (possible) reason we could mention might be compiled code size. I assume we don't have a desire to compile without this, maybe for mobile?

nairbv commented:
> There will be followups.

I assume this means there are places in the code where the flag will still be present after this PR? What are those, and then happens if we try to compile with it turned off?

zou3519 commented:
> One other (possible) reason we could mention might be compiled code size. I assume we don't have a desire to compile without this, maybe for mobile?

I can note that, but we don't have a desire to compile without named tensors. All builds right now (including mobile) have named tensors on. 

> > There will be followups.
> 
> I assume this means there are places in the code where the flag will still be present after this PR? What are those[?]

The last cases are deleting BUILD_NAMEDTENSOR from python code generation. I don't want to include everything in this PR because that might result in lots of merge conflicts. In particular, I might even split this PR up to avoid merge conflicts.

> and then what happens if we try to build with it turned off?

After this PR, we won't be able to build with named tensors off -- that will be an unsupported operation.

nairbv commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -53,9 +53,7 @@
 #include <torch/csrc/jit/tracer.h>
 #include <torch/csrc/jit/ir.h>
 #include <ATen/core/EnableNamedTensor.h>
```
should we be excluding EnableNamedTensor too?

zou3519 committed:
Update on "Remove BUILD_NAMEDTENSOR macros"


This PR begins the process of removing BUILD_NAMEDTENSOR macros. There
will be followups.

Reasons for removing the macros:
- BUILD_NAMEDTENSOR is always on and has been on since pytorch 1.3.0.
- Since we don't test building without it, it is useless to keep around.
- Code becomes nicer to read without the macros

Reasons for not removing the macros:
- potential for feature flagging
- binary size reduction / pytorch modularity. I don't think this is a
major consideration because macros like this are not the best way to
achieve this.

Now, I argue against needing to feature flag. The main reason why we
might want to feature flag is if we need to disable the feature.
We'd need a fast switch to disable the feature if someone discovers
in the future that named tensors caused some regression in some existing workflows.

In https://github.com/pytorch/pytorch/pull/25798, I did a variety of
macro- and micro- benchmarks to determine the performance impact of named
tensors on regular tensors.

[The
microbenchmarks](https://github.com/pytorch/pytorch/pull/25798#issuecomment-529014810)
were not very stable, and running the
microbenchmarks for more iterations doesn't actually help because the
noise is not distributed in a nice way. Instead of microbenchmarks I ran
a [profiler
(perf)](https://github.com/pytorch/pytorch/pull/25798#issuecomment-555707645)
to estimate how much overhead named tensors add to unnamed code. I
estimated the overhead to be less than 100ns for `add` and even smaller
for `mm`; there are ways to optimize even futher if we find this to be a
problem.

[Initial
macrobenchmarks](https://github.com/pytorch/pytorch/pull/25798#issuecomment-530539104)
were also not very stable. I ran imagenet for some number of epochs. To
make them more stable, I got rid of the data loading (which seemed to
vary between runs). [In some benchmarkers without data
loading](https://github.com/pytorch/pytorch/pull/25798#issuecomment-562214053),
we can see that the results are less noisy now. These results support
no noticeable regressions in speed.

Test Plan:
- wait for CI

Differential Revision: [D18858543](https://our.internmc.facebook.com/intern/diff/D18858543)

[ghstack-poisoned]
